### DISCUSSION TITLE
How to find modbus address location

### DISCUSSION THREAD

**Discussion by JshStadler** [syssi/esphome-smg-ii#30] [Q&A]
I'm running into the below problem, and I suspect it's because of different addresses being used by my inverter It's an RCT Axpert unit, but I was hoping someone might be able to tell me how I can determine what addresses is being used by my inverter so I can try to update it in the YAML to hopefully get it working for my inverter. 
The pipsolar ESPHome integration does work for my inverter, however it doesn't allow me to change the charging current and bulk/float voltages, so I was hoping I might be able to do so using this: https://github.com/syssi/esphome-smg-ii/blob/main/esp32-example.yaml

`Duplicate modbus command found: type 0x3 address:XXX`

```
14:07:04 | [D] | [uart_debug:114] | >>> 01:03:00:EA:00:01:A5:FE
14:07:04 | [D] | [modbus_controller:032] | Modbus command to device=1 register=0xEA countdown=0 no response received - removed from send queue
14:07:05 | [D] | [uart_debug:114] | >>> 01:03:01:2D:00:03:94:3E
14:07:05 | [W] | [modbus_controller:113] | Duplicate modbus command found: type=0x3 address=301 count=3
14:07:05 | [W] | [modbus_controller:113] | Duplicate modbus command found: type=0x3 address=313 count=1
14:07:05 | [W] | [modbus_controller:113] | Duplicate modbus command found: type=0x3 address=320 count=2
14:07:05 | [D] | [uart_debug:114] | >>> 01:03:01:2D:00:03:94:3E
14:07:05 | [D] | [uart_debug:114] | >>> 01:03:01:2D:00:03:94:3E
14:07:05 | [D] | [uart_debug:114] | >>> 01:03:01:2D:00:03:94:3E
```
![image](https://github.com/syssi/esphome-smg-ii/assets/37085670/7ccc8395-268e-45e1-bbd7-1fdd159866d4)

**Comment by syssi** [comment]
Do you own any OEM software (a WiFi dongle, a windows application or something else) to capture valid requests?

**Reply by JshStadler** [reply]
Hi syssi, thanks for the quick response.

I'm not sure OEM software you are referring to, but I do have a USB to Serial adaptor which I can use with MultiSIBControl to monitor the inverter and adjust settings. (If needed I can also setup a Lunix machine with SolPipLog)

As for the stuff I used to test your modbus setup, I had the inverter connected to an ESP32 via a MAX3232 module.

**Reply by syssi** [reply]
Did you understand my idea? If there is any other software which is able to control your inverter properly we should be able to sniff the frames / communication and extract the proper register addresses.

**Reply by JshStadler** [reply]
I now understand your idea :) 
Would I need a second USB to serial adaptor, or can I use an ESP32 with a MAX3232 converter to do the sniffing? If I understand correctly, the idea is to capture the communication data that is transmitted and received over serial when a change is made to the desired settings, which should have the target address of the setting which I can make use of in your YAML config for ESPHome.

I'm keen to give it a try, but I'm a bit clueless on where to start. Do you have a guide or specific hardware you'd recommend that I can make use of?

**Reply by syssi** [reply]
Let's identify a software first which is compatible with your inverter so you are able to control the device in general. Please focus on the commands which doesn't work at the moment. Could you try different applications you are aware of?

**Reply by JshStadler** [reply]
Adjusting the settings through this program works for my inverter: http://www.multisibcontrol.net/

Watchpower is also supposed to work if that's easier (I just never used it as MultiSIBControl seemed to work with less hassle).
I have a Mecer USB to 9-pin Serial AP1103 adaptor that I use to connect the inverter to the computer.

(I also have the latest inverter firmware files which is the current version of the firmware I have on the inverter)

**Reply by syssi** [reply]
Watchpower has a debug tab. So you can copy&paste the raw commands from there. No complicated sniffing/capturing required.

**Reply by syssi** [reply]
MutliSIBControl has a `Console` tab which shows the raw traffic too: http://www.multisibcontrol.net/images/s9m.jpg

**Reply by JshStadler** [reply]
Hi syssi,
Thanks so much for the help, I will try to get the addresses from there.

I seem to have run into an unfortunate delay. I was struggling to get the inverter recognised in MultiSIBControl through the USB to Serial adaptor (I might be because it was a different laptop than I previously used to connect to the inverter, and the serial drivers was meant for Arduino/ESP).
But right after I uninstalled the Serial drivers for the adaptor (to install the correct ones), I heard a slight popping sound from inside the inverter, and now the communications port isn't working (the ESP is also not getting any data from it now). I guess the adaptor can be dangerous while there is no driver installed. Luckily, my inverter is still under warranty, so it should be covered as I doubt this is user error.

Once I get the communications board replaced, I will reattempt to get the addresses, and provide updates here.

**Reply by JshStadler** [reply]
So, I've managed to get the communication board replaced and it's working now. MultiSIBControl doesn't seem to be able to change either the bulk/float charge voltages or charging current limits (they just fail to change, everything else seem to work except equalisation frequency), but the latest version of WatchPower from Voltronics is able to adjust the charging current, but still not the float voltages.

MultiSIBControl shows that it's using these protocols in the logs: MUCHGC (Max Utility Charging Current), MCHGC (Max Charging Current), PBFT (Float Voltage), PCVV (Bulk Voltage).
As for WatchPower, I can't seem to find a console/log view that has to do with the protocols or addresses that's being used/targeted.

**Reply by syssi** [reply]
It's a bit hard for me to understand what's not working if you use the `pipsolar` component and what's working if you use one of the windows applications. Could you try to provide a comparison of the different commands/payloads? Please keep in mind we are talking about a different protocol at the moment. The PIP protocol is text based vs binary & structured modbus frames.

**Reply by JshStadler** [reply]
Hi @syssi 
Sorry for only responding now.

I'm not sure how to get to the modbus commands/payloads in either WatchPower or MultiSIBControl. I was only able to use the commands from the attached document, as from my understanding, I thought I'd need to use those commands and then I might be able to see what address values are being changed. 

Do you think something like SolPipLog would be better to try? Also, if I'm not supposed to use modbus commands, are there specific commands I should use?

[HS_MS_MSX_RS232_Protocol_20140822_after_current_upgrade.pdf](https://github.com/syssi/esphome-smg-ii/files/11851470/HS_MS_MSX_RS232_Protocol_20140822_after_current_upgrade.pdf)

**Comment by JshStadler** [comment]
I'm sorry for the confusion, I'm just referring to the `pipsolar` protocols as those are the ones that I was able to see being used by MultiSIBUtil (I have no idea what WatchPower uses). 

This is the version of WatchPower I used: http://ftps.voltronic.com.tw/V/ViewPower-C76A89B88C10454AA513/WatchPower/WatchPower1.16SP9/Windows/

WatchPower doesn't seem be have a place where I can see useful logs or enter commands. It is however able to adjust the Max Charging Current (which doesn't work on MultiSIBUtil), but neither of them are able to adjust the float/bulk charge voltages (I'll hear from the supplier of the inverter if they can test on their side why the program the inverter recommends can't adjust the charge voltages).
WatchPower did however have an option to "debug" but that only generated the below logs.
[WatchPower Debug Logs.zip](https://github.com/syssi/esphome-smg-ii/files/11712865/WatchPower.Debug.Logs.zip)

Here's a link to videos where I made changes to the settings and attempted to view the logs or make use of the commands, it should give you an overview of what I was able to do in both applications (the firmware update files is also in there if it helps at all):
https://drive.google.com/drive/folders/1BupDJ9RBN3cl4s3gHAHWs2M9z6491XTL?usp=sharing

Can I ask how you managed to get the address values for adjusting the settings? Maybe it's easier if I just try to do the same as what you did.

**Reply by syssi** [reply]
If you take a look at the `docs` directory of the project you will find a protocol description PDF. I've just implemented the mentioned register addresses.
